scripts/GasEstimation.ts

```diff
@@ -0,0 +1,435 @@
+    )
+  ).to.equal(true, 'Worker did not receive payment');
+}
+
+function correctEstimatedCallCost(estimation: BigNumber) {
+  const internalCorrection = BigNumber.from(
+    INTERNAL_TRANSACTION_ESTIMATED_CORRECTION
+  );
+  const gasCorrectionFactor = BigNumber.from(ESTIMATED_GAS_CORRECTION_FACTOR);
+
+  return gasCorrectionFactor.mul(
+    estimation.gt(internalCorrection)
+      ? estimation.sub(internalCorrection)
+      : estimation
+  );
+}
+
+async function getCurrentBalances(
+  token: UtilToken,
+  forwarder: SupportedSmartWallet
+) {
+  const forwarderBalance = await token.balanceOf(forwarder.address);
+  const relayWorkerBalance = await token.balanceOf(relayWorker.address);
+
+  return {
+    forwarderBalance,
+    relayWorkerBalance,
+  };
+}
+
+async function getEstimatedGasWithCorrection(
+  senderAddress: string,
+  receiverAddress: string,
+  data: string
+) {
+  const estimatedDestinationCallGas = await ethers.provider.estimateGas({
+    from: senderAddress,
+    to: receiverAddress,
+    gasPrice: GAS_PRICE,
+    data,
+  });
+
+  return correctEstimatedCallCost(estimatedDestinationCallGas);
+}
+
+async function completeRelayRequest(
+  transferReceiver: string,
+  balanceToTransfer: string,
+  fees: string,
+  token: UtilToken,
+  relayRequest: RelayRequest,
+  forwarder: SmartWallet
+) {
+  const isSponsored = fees === '0';
+  const tokenContract = isSponsored
+    ? ethers.constants.AddressZero
+    : token.address;
+
+  const encodedFunction = token.interface.encodeFunctionData('transfer', [
+    transferReceiver,
+    balanceToTransfer,
+  ]);
+
+  const estimatedDestinationCallGasCorrected =
+    await getEstimatedGasWithCorrection(
+      forwarder.address,
+      token.address,
+      encodedFunction
+    );
+
+  let tokenGas = '0';
+
+  if (!isSponsored) {
+    const dataForTransfer = token.interface.encodeFunctionData('transfer', [
+      relayWorker.address,
+      fees,
+    ]);
+
+    tokenGas = (
+      await getEstimatedGasWithCorrection(
+        forwarder.address,
+        token.address,
+        dataForTransfer
+      )
+    ).toString();
+  }
+
+  return combineTwoRelayRequests(relayRequest, {
+    request: {
+      data: encodedFunction,
+      to: token.address,
+      nonce: (await forwarder.nonce()).toString(),
+      tokenAmount: fees,
+      tokenContract,
+      gas: estimatedDestinationCallGasCorrected.toString(),
+      tokenGas,
+    },
+  });
+}
+
+async function estimateRelayCost(fees = '0') {
+  const recipient = await deployContract<TestRecipient>('TestRecipient');
+  const { relayHub, smartWalletFactory, owner, token, verifier } =
+    await deployAndSetup();
+
+  const smartWallet = (await createSupportedSmartWallet({
+    relayHub: relayHubSigner.address,
+    factory: smartWalletFactory,
+    owner,
+    sender: relayHubSigner,
+  })) as SmartWallet;
+
+  const baseRelayRequest = {
+    request: {
+      relayHub: relayHub.address,
+      to: recipient.address,
+      data: '0xdeadbeef',
```

Comment:
what is this data?

Reply:
Just dummy data. Dead beef, probably some meat lover wrote it.

Reply:
I removed the recipient and the data related to the call.